k2/csrc/determinize.cc

```diff
@@ -5,469 +5,287 @@
+  std::unordered_set<TracebackState *> cur_states;
+
+  double fb_prob = -std::numeric_limits<double>::infinity();
+  for (auto p : elements) {
+    TracebackState *state = p.second.get();
+    fb_prob = LogSumOrMax<TracebackState>(
+        fb_prob,
+        state->forward_prob + wfsa_in.backward_state_weights[state->state_d]);
+    cur_states.insert(state);
   }
 
-  struct DetStateHasher {
-    size_t operator()(const std::pair<uint64_t, uint64_t> &p) const {
-      return p.first;
-    }
-  };
-};
+  int32_t new_seq_len = GetMostRecentCommonAncestor(&cur_states);
+  // now cur_states.size() == 1.
+  CHECK_EQ(cur_states.size(), 1);
+  CHECK_LE(new_seq_len, seq_len);
+
+  const TracebackState *base_state = cur_states.front().get();
+  // The following statement is a correction term that we add to
+  // forward_backward_prob, in which we replace the forward_prob in the DetState
+  // (which will have been computed in a path-dependent way) with the
+  // forward_prob in wfsa_in.  Note: the values of state->forward_prob above can
+  // be thought of as base_state->forward_prob plus some value that only depends
+  // on the symbol sequence.  The point of this is to ensure that
+  // this->forward_backward_prob (which is used for pruning) depends only on the
+  // base_state and the symbol sequence, and not on "how we got here", i.e.  the
+  // history of DetStates from which this one is derived via ProcessArcs().
+  fb_prob += wfsa_in.forward_state_weights[base_state->state_id] -
+             base_state->forward_prob;
+  // set thi->forward_backward_prob; it will affect pruning.
+  this->forward_backward_prob = fb_prob;
+  this->seq_len = new_seq_len;
+
+  // the following will set removed_weight and deriv_info.
+  TraceBack(&cur_states, seq_len - new_seq_len, wfsa_in.arc_weights,
+            removed_weight, deriv_info);
+
+  normalized = true;
+}
 
-void DeterminizeMax(const WfsaWithFbWeights &a, float beam, Fsa *b,
-                    std::vector<std::vector<int32_t>> *arc_map) {
-  // TODO(dpovey): use glog stuff.
-  assert(IsValid(a) && IsEpsilonFree(a) && IsTopSortedAndAcyclic(a));
-  if (a.arc_indexes.empty()) {
-    b->Clear();
-    return;
+template <typename TracebackState>
+float DeterminizePrunedTpl(
+    const WfsaWithFbWeights &wfsa_in, float beam, int64_t max_step,
+    Fsa *fsa_out, std::vector<float> *arc_weights_out,
+    std::vector<std::vector<typename TracebackState::DerivType>>
+        *arc_derivs_out) {
+  CHECK_GT(beam, 0);
+  CHECK(IsDeterministic(wfsa_in.fsa));
+  CHECK(!IsEmpty(wfsa_in.fsa));
+
+  DetStatePriorityQueue<TracebackState> queue;
+  DetStateMap<TracebackState> map;
+  using DS = DetState<TracebackState>;
+
+  std::shared_ptr<DS> start_state = std::make_shared<DS>();
+
+  std::vector<Arc> arcs_out;
+  arc_weights_out->clear();
+  arc_derivs_out->clear();
+
+  bool ans = map.GetOutputState(start_state.get());
+  CHECK(ans && ans->state_id == 0);
```

Comment:
again, seems we should change the signature of `GetOutputState`?

Reply:
Yes I think so (add wfsa_in.fsa)